### PR TITLE
fix: layer button component bugs

### PR DESCRIPTION
**Pull request by pramodcog**
#### Type of change
Bug 

#### Jira ticket :blue_book:

https://cognitedata.atlassian.net/browse/UX-1126

## Description :pencil:
Bug fixes listed in UX-1126 and some more edge cases handled

https://github.com/cognitedata/reveal/assets/87521752/2e251bb5-5dc7-466e-b98d-b93fdd7c08f7

## How has this been tested? :mag:

In Storybook

## Test instructions :information_source:

`cd react-components && yarn && yarn storybook`
Select `LayersContainer` example

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [ ] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

**Review by christjt — CHANGES_REQUESTED** [review]
Bug1: Camera resets when toggling 360 !?!?!?

Bug2: Initial time when opening the side panel a double click is needed to select / deselect

Bug3 (from your video): There is something wrong when hovering 
image

Probably bug: Fit camera to models button should only fit to visible models.

File: react-components/src/components/RevealToolbar/LayersButton.tsx [hunk 1]
```diff
@@ -2,29 +2,121 @@
...
+  const updatedReveal3DResources: Reveal3DResourcesStates = useMemo(() => {
+    if (cadModelName.data === null && pointCloudModelName.data === null) {
+      return {
+        cadModels: [],
+        pointCloudModels: [],
+        image360Collections: []
+      };
+    }
+    const cadModels = viewer.models.filter((model) => model.type === 'cad');
+    const pointCloudModels = viewer.models.filter((model) => model.type === 'pointcloud');
```
Comment (christjt) [inline react-components/src/components/RevealToolbar/LayersButton.tsx:71]: You can use a type guard here such that you don't need to cast later.

File: react-components/src/components/RevealToolbar/LayersContainer/PointCloudLayersContainer.tsx [hunk 2]
```diff
@@ -79,19 +72,24 @@ export const PointCloudLayersContainer = (): ReactElement => {
...
+      {pointCloudModels.length > 0 && (
+        <Menu.Submenu openOnHover={false} content={pointCloudModelContent()} title="Point clouds">
+          <Flex direction="row" justifyContent="space-between">
+            <Checkbox
+              key={`allPointCLoudModelCheckbox-${String(allModelVisible)}-${String(indeterminate)}`}
+              checked={allModelVisible}
+              indeterminate={indeterminate}
+              onChange={(e, c) => {
+                e.stopPropagation();
+                handleAllPointCloudModelsVisibility(c as boolean);
```
Comment (christjt) [inline react-components/src/components/RevealToolbar/LayersContainer/PointCloudLayersContainer.tsx:85]: Why is `c` not typed here?

File: react-components/src/components/RevealToolbar/LayersContainer/types.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+import type React from 'react';
```
Comment (christjt) [inline react-components/src/components/RevealToolbar/LayersContainer/types.ts:4]: Any reason why you import default React?

File: react-components/src/components/RevealToolbar/LayersButton.tsx [hunk 4]
```diff
@@ -2,29 +2,121 @@
...
+import {
+  type CognitePointCloudModel,
+  type CogniteCadModel,
+  type CogniteModel
+} from '@cognite/reveal';
+import { useReveal } from '../RevealContainer/RevealContext';
+import { use3DModelName } from '../../hooks/use3DModelName';
+import isEqual from 'lodash/isEqual';
 
 export const LayersButton = (): ReactElement => {
```
Comment (christjt) [inline react-components/src/components/RevealToolbar/LayersButton.tsx:18]: Try to refactor this component into custom hooks / local functions / lambdas. Very hard to follow the logic here.

File: react-components/src/components/RevealToolbar/LayersContainer/elements.ts [hunk 5]
```diff
@@ -16,6 +16,25 @@ export const StyledMenu = styled(Menu)`
 export const StyledSubMenu = styled(Menu)`
   box-shadow: none;
   padding: 8px;
+  overflow-y: auto;
+  max-height: 300px;
+  ::-webkit-scrollbar {
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply react-components/src/components/RevealToolbar/LayersContainer/elements.ts:21]: `::-webkit` styles are used in `fusion` front-end across apps such as FDX & Industry-canvas, ref: https://github.com/cognitedata/fusion/blob/7f4a21732cfcb821d64b9d854adc57e6ce435255/apps/flexible-data-explorer/src/app/containers/Filter/components/MenuList/elements.ts#L7

File: react-components/src/components/RevealToolbar/LayersContainer/CadModelLayersContainer.tsx [hunk 6]
```diff
@@ -2,38 +2,29 @@
...
-      name: modelName?.data?.[index] ?? 'No model name'
-    }))
-  );
-
-  const [allCadModelVisible, setAllCadModelVisible] = useState(true);
-  const [indeterminate, setIndeterminate] = useState<boolean>(false);
-
-  const count = selectedCadModels.length.toString();
+  const count = cadModels.length.toString();
+  const allModelVisible = !cadModels.every((data) => !data.isToggled);
```
Comment (haakonflatval-cognite) [inline react-components/src/components/RevealToolbar/LayersContainer/CadModelLayersContainer.tsx:23]: Not sure I follow the logic of this - `allModelVisible` is the same as `not every cadmodel is not toggled`? Same with the other model types too

**Review by haakonflatval-cognite — CHANGES_REQUESTED** [review]
Looks like Christopher already gave you a big list of things to do :sweat_smile: I'll just add some nitpick:

- When inside a 360 image and toggling only that 360 set, the viewer does not exit that image
- Still feel like it should be possible to deduplicate the code across layer buttons :thinking: Something along the lines of wrapping every model in types implementing a common interface to get/set toggle etc. But not necessary for this PR

File: react-components/src/components/RevealToolbar/LayersContainer/PointCloudLayersContainer.tsx (hunk 2 above)
Comment (pramodcog) [reply react-components/src/components/RevealToolbar/LayersContainer/PointCloudLayersContainer.tsx:85]: `c` in `onChange` is `string | boolean | undefined`

**Review by haakonflatval-cognite — DISMISSED** [review]

**Comment by pramodcog** [comment]
> Bug1: Camera resets when toggling 360 !?!?!?
> 
> Bug2: Initial time when opening the side panel a double click is needed to select / deselect
> 
> Bug3 (from your video): There is something wrong when hovering image
> 
> Probably bug: Fit camera to models button should only fit to visible models.

Bug 2 & 3 are related to cogs.js library, have connected & communicated to the team

File: react-components/src/components/RevealToolbar/LayersContainer/Image360LayersContainer.tsx [hunk 7]
```diff
@@ -74,19 +77,24 @@ export const Image360CollectionLayerContainer = (): ReactElement => {
...
-        <StyledLabel> 360 images </StyledLabel>
-        <StyledChipCount label={count} hideTooltip type="neutral" />
-      </Flex>
-    </Menu.Submenu>
+    <>
+      {image360LayerData.length > 0 && (
+        <Menu.Submenu content={image360Content()} title="360 images">
+          <Flex direction="row" justifyContent="space-between">
+            <Checkbox
+              key={`all360ImagesCheckbox-${String(someImagesVisible)}-${String(indeterminate)}`}
```
Comment (haakonflatval-cognite) [inline react-components/src/components/RevealToolbar/LayersContainer/Image360LayersContainer.tsx:85]: Do we need a `key` property here? Isn't that only necessary when we have multiple similar elements inside the same parent?

**Review by haakonflatval-cognite — APPROVED** [review]
Looks good to me! Apart from the cogs.js-issues, at least :P

File: react-components/src/components/RevealToolbar/LayersContainer/Image360LayersContainer.tsx (hunk 7 above)
Comment (pramodcog) [reply react-components/src/components/RevealToolbar/LayersContainer/Image360LayersContainer.tsx:85]: Checkbox needs to trigger re-render when either `someImagesVisible` or `indeterminate` value changes, so when `key` changes, it would trigger the re-render on the component

File: react-components/src/components/RevealToolbar/LayersContainer/PointCloudLayersContainer.tsx (hunk 2 above)
Comment (christjt) [reply react-components/src/components/RevealToolbar/LayersContainer/PointCloudLayersContainer.tsx:87]: Ok, so what happens if this is actually a string then?

File: react-components/src/components/RevealToolbar/LayersContainer/types.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
```
Comment (christjt) [inline react-components/src/components/RevealToolbar/LayersContainer/types.ts:4]: Its still here 😅

File: react-components/src/components/RevealToolbar/LayersContainer/Image360LayersContainer.tsx [hunk 9]
```diff
@@ -74,19 +77,23 @@ export const Image360CollectionLayerContainer = (): ReactElement => {
...
-      </Flex>
-    </Menu.Submenu>
+    <>
+      {image360LayerData.length > 0 && (
+        <Menu.Submenu content={image360Content()} title="360 images">
+          <Flex direction="row" justifyContent="space-between">
+            <Checkbox
+              checked={someImagesVisible}
+              indeterminate={indeterminate}
+              onChange={(e, c) => {
```
Comment (christjt) [inline react-components/src/components/RevealToolbar/LayersContainer/Image360LayersContainer.tsx:87]: What happens if this is string?

**Review by christjt — CHANGES_REQUESTED** [review]
Some more minor comments

**Review by christjt — APPROVED** [review]
I still have the "needing to double click" first time issue. But I'll assume you will follow this up. Other than that, it looks good.

**Comment by pramodcog** [comment]
> I still have the "needing to double click" first time issue. But I'll assume you will follow this up. Other than that, it looks good.

It is bug in cogs.js, the CDS team have acknowledge the bug, will follow up on it

**Review by haakonflatval-cognite — APPROVED** [review]
LGTM :+1:
